In [68]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [69]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [70]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [71]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [72]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
normalized_data = scaler.fit_transform(df_market_data)
normalized_data = pd.DataFrame(normalized_data, columns=df_market_data.columns)
normalized_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
0                     0.508529                    0.493193   
1                     0.185446                    0.934445   
2                     0.021774                   -0.706337   
3                    -0.040764                   -0.810928   
4                     1.193036                    2.000959   

   price_change_percentage_14d  price_change_percentage_30d  \
0                     0.772200                     0.235460   
1                     0.558692                    -0.054341   
2                    -0.021680                    -0.061030   
3                     0.249458                    -0.050388   
4                     1.760610                     0.545842   

   price_change_percentage_60d  price_change_percentage_200d  \
0                    -0.067495                     -0.355953   
1                    -0.273483                     -0.115759   
2                     0.008005                     -0.550247   
3                    -0.373164                     -0.458259   
4                    -0.291203                     -0.499848   

   price_change_percentage_1y  
0                   -0.251637  
1                   -0.199352  
2                   -0.282061  
3                   -0.295546  
4                   -0.270317

In [73]:
# Create a DataFrame with the scaled data
scaled_df = pd.DataFrame(normalized_data, columns=df_market_data.columns)

# # Copy the crypto names from the original data
scaled_df["coin_id"] = df_market_data.index

# # Set the coinid column as index
scaled_df.set_index('coin_id', inplace=True)

# # Display sample data
scaled_df.sample(25)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
usd-coin                                 0.034352                   -0.733026   
true-usd                                 0.078038                   -0.687745   
chainlink                                0.011397                    2.572251   
tron                                     0.130050                   -0.041018   
theta-token                             -1.612188                   -1.682027   
nem                                     -0.217984                   -0.849381   
ontology                                -0.409044                   -0.906963   
neo                                      0.286546                   -0.326301   
binancecoin                              0.891871                    1.327295   
wrapped-bitcoin                          0.515453                    0.461843   
tether                                   0.021774                   -0.706337   
eos                                      0.155710                   -0.922491   
ethlend                                 -4.981042                   -0.045178   
vechain                                  0.585089                   -0.994231   
cardano                                  0.102530                    1.508001   
crypto-com-chain                         0.331280                   -1.614844   
ethereum                                 0.185446                    0.934445   
okb                                     -0.923203                   -1.437359   
ftx-token                                0.414711                    0.414044   
iota                                     0.259097                    0.249508   
bitcoin                                  0.508529                    0.493193   
celsius-degree-token                     1.045530                   -0.618328   
zcash                                   -0.127467                    0.929119   
cdai                                     0.180851                   -0.704931   
ethereum-classic                        -0.071312                   -0.229484   

                      price_change_percentage_14d  \
coin_id                                             
usd-coin                                -0.023140   
true-usd                                -0.009191   
chainlink                                1.101647   
tron                                     0.147155   
theta-token                             -0.816921   
nem                                      0.297632   
ontology                                -1.298986   
neo                                     -1.212670   
binancecoin                              0.800214   
wrapped-bitcoin                          0.769975   
tether                                  -0.021680   
eos                                      0.115024   
ethlend                                 -1.206956   
vechain                                 -2.217108   
cardano                                  0.648885   
crypto-com-chain                        -1.054521   
ethereum                                 0.558692   
okb                                     -0.629963   
ftx-token                               -0.047386   
iota                                    -0.478953   
bitcoin                                  0.772200   
celsius-degree-token                     2.907054   
zcash                                    0.677532   
cdai                                    -0.001816   
ethereum-classic                        -0.175544   

                      price_change_percentage_30d  \
coin_id                                             
usd-coin                                -0.065775   
true-usd                                -0.058214   
chainlink                               -0.490495   
tron                                    -0.543776   
theta-token                              1.148607   
nem                            

---

### Find the Best Value for k Using the Original Data.

In [74]:
# Create a list with the number of k-values from 1 to 11 
# *** In all of the examples we used in class this was a k = list(range(1,11)) Using this as youre asking specifically for a list from 1-11.
k = list(range(1, 12))
print(k)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [75]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(scaled_df)
    inertia.append(k_model.inertia_)

f:\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [76]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

df_elbow.head()

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.405923

In [77]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
k_elbow_plot = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

# Plot K elbow
k_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for k looks like 4 based on the elbow curve.


---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [78]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [79]:
# Fit the K-Means model using the scaled data
model.fit(scaled_df)

KMeans(n_clusters=4, random_state=1)

In [80]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_lower = model.predict(scaled_df)

# Print the resulting array of cluster values.
print(k_lower)

[3 3 1 1 3 3 3 3 3 1 1 1 1 3 1 3 1 1 3 1 1 3 1 1 1 1 1 1 3 1 1 1 0 3 1 1 2
 1 1 1 1]


In [81]:
# Create a copy of the DataFrame
copy_of_scaled_df = scaled_df.copy()

In [82]:
# Add a new column to the DataFrame with the predicted clusters
copy_of_scaled_df['clusters_lower'] = k_lower

# Display sample data
copy_of_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters_lower  
coin_id                                                   
bitcoin                        -0.251637               3  
ethereum                       -0.199352               3  
tether                         -0.282061               1  
ripple                         -0.295546               1  
bitcoin-cash                   -0.270317               3

In [83]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
scaled_plot = copy_of_scaled_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols="coin_id",
    by="clusters_lower",
).opts(yformatter="%.0f")

#Plot Scaled Plot
scaled_plot


:NdOverlay   [clusters_lower]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [84]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [85]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
ccinfo_pca = pca.fit_transform(scaled_df)
# View the first five rows of the DataFrame. 
ccinfo_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [86]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** = 0.3719856 + 0.34700813 + 0.17603793 = 0.89503166
Almost 90% from 3 different principal components.

In [87]:
# Create a new DataFrame with the PCA data.
scaled_df_pca = pd.DataFrame(ccinfo_pca, columns=["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d"])
# Copy the crypto names from the original data
crypto = df_market_data.index

# Set the coinid column as index
scaled_df_pca.index = crypto

# Display sample data
scaled_df_pca.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                         -0.600667                    0.842760   
ethereum                        -0.458261                    0.458466   
tether                          -0.433070                   -0.168126   
ripple                          -0.471835                   -0.222660   
bitcoin-cash                    -1.157800                    2.041209   

              price_change_percentage_14d  
coin_id                                    
bitcoin                          0.461595  
ethereum                         0.952877  
tether                          -0.641752  
ripple                          -0.479053  
bitcoin-cash                     1.859715

---

### Find the Best Value for k Using the PCA Data

In [88]:
# Create a list with the number of k-values from 1 to 11 
# *** In all of the examples we used in class this was a k = list(range(1,11)) Using this as youre asking specifically for a list from 1-11.
k = list(range(1, 12))
print(k)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [89]:
# Create an empy list to store the inertia values
inertia = []
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(scaled_df_pca)
    inertia.append(model.inertia_)

f:\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [90]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k": k,
    "inertia": inertia
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca.head()

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [91]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
pca_elbow_plot = df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve using PCA", 
    xticks=k
)
# Plot PCA elbow
pca_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** According to the PCA is looks like k=4 has the last noticable bend in the curve.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The Data does not differ as K=4 in the original data and here it looks like K=4 as well.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [92]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [93]:
# Fit the K-Means model using the PCA data
model.fit(scaled_df_pca)

KMeans(n_clusters=4, random_state=1)

In [94]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_3 = model.predict(scaled_df_pca)
# Print the resulting array of cluster values.
k_3

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [95]:
# Create a copy of the DataFrame with the PCA data
pca_predictions_df = scaled_df_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_predictions_df["predicted_clusters"] = k_3

# Display sample data
pca_predictions_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                         -0.600667                    0.842760   
ethereum                        -0.458261                    0.458466   
tether                          -0.433070                   -0.168126   
ripple                          -0.471835                   -0.222660   
bitcoin-cash                    -1.157800                    2.041209   

              price_change_percentage_14d  predicted_clusters  
coin_id                                                        
bitcoin                          0.461595                   1  
ethereum                         0.952877                   1  
tether                          -0.641752                   0  
ripple                          -0.479053                   0  
bitcoin-cash                     1.859715                   1

In [96]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_plot = pca_predictions_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols="coin_id",
    by="predicted_clusters"
)
# Plot PCA
pca_plot

:NdOverlay   [predicted_clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [99]:
# Composite plot to contrast the Elbow curves
composite_elbow = k_elbow_plot * pca_elbow_plot
composite_elbow

:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [98]:
# Composite plot to contrast the clusters
composite_clusters = scaled_plot + pca_plot
composite_clusters

:Layout
   .NdOverlay.I  :NdOverlay   [clusters_lower]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predicted_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features gets us a very similar end result to the initial model. We can identify the clusters in both. The loss of information is about 10% but the information still portrays the groupings well. The second model looks a little bit tighter but both are very similar. The impact on ethlend & celsius-degree-token was the most noticable between the two charts.